[Merged by Bors] - feat(tactic/dependencies): add tactics to compute (reverse) dependencies

JLimperg opened the pull request:
These are the beginnings of an API about dependencies between expressions. For
now, we only deal with dependencies and reverse dependencies of hypotheses,
which are easy to compute.

Breaking change: `tactic.revert_deps` is renamed to
`tactic.revert_reverse_dependencies_of_hyp` for consistency. Its implementation
is completely new, but should be equivalent to the old one except for the order
in which hypotheses are reverted. (But the old implementation made no particular
guarantees about this order anyway.)

JLimperg committed:
feat(tactic/dependencies): add tactics to compute (reverse) dependencies of hyps

These are the beginnings of an API about dependencies between expressions. For
now, we only deal with dependencies and reverse dependencies of hypotheses,
which are easy to compute.

Breaking change: `tactic.revert_deps` is renamed to
`tactic.revert_reverse_dependencies_of_hyp` for consistency. Its implementation
is completely new, but should be equivalent to the old one except for the order
in which hypotheses are reverted. (But the old implementation made no particular
guarantees about this order anyway.)

eric-wieser commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,230 @@
+    local_def_value_has_local_in h ns ]
+
+/--
+Given a hypothesis `h`, `hyp_depends_on_locals h ns` returns true iff `h`
```
```suggestion
Given a hypothesis `h`, `hyp_depends_on_locals_inclusive h ns` returns true iff `h`
```

robertylewis commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,230 @@
+
+/--
+Given a hypothesis `h`, `hyp_depends_on_locals h ns` returns true iff `h`
+depends on a local constant whose unique name appears in `ns`. See note
```
```suggestion
depends on a local constant whose unique name appears in `ns`. See Note
```
and below. I think the regex parser in doc gen only looks for capital `Note [...]`.

robertylewis commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,230 @@
+
+/--
+Given a hypothesis `h` and local constant `i`, `hyp_depends_on_local h i`
+checks whether `h` depends on `i`. See note [dependencies of hypotheses].
```
What is the behavior if `i` is not a local constant?

robertylewis commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,230 @@
+  pure $ deps.insert h.local_uniq_name
+
+/--
+Given a hypothesis `h`, `dependencies_of_hyp_inclusive' h` returns the
```
```suggestion
Given a hypothesis `h`, `dependencies_of_hyp_inclusive h` returns the
```

robertylewis commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,230 @@
+hyp_depends_on_locals h (mk_name_set.insert i.local_uniq_name)
+
+/--
+Given a hypothesis `h` and local constant `i`, `hyp_depends_on_local h i`
```
```suggestion
Given a hypothesis `h` and local constant `i`, `hyp_depends_on_local_inclusive h i`
```

robertylewis commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,230 @@
+  hs.foldl (λ ns h, ns.insert h.local_uniq_name) mk_name_set
+
+/--
+Revert the local constants whose unique names appear in `hs`, as well as any
```
```suggestion
`revert_set hs` reverts the local constants whose unique names appear in `hs`, as well as any
```
and below

robertylewis reviewed (commented):
I think these additions look very good!

JLimperg committed:
Make 'depends on' relation transitive when there are local defs

JLimperg committed:
Add tests

JLimperg committed:
Placate the linter

JLimperg committed:
Add variants for list expr, name_set and expr_set for all functions

I thought these trivial variants would just clutter the API, but then I
immediately needed a couple of them for another tactic.

robertylewis commented on src/meta/rb_map.lean:
```diff
@@ -200,3 +200,12 @@ meta instance {data : Type} : inhabited (name_map data) :=
+
+namespace expr_set
+
+meta def local_set_to_name_set (lcs : expr_set) : name_set :=
```
missing doc string

JLimperg committed:
Fix type errors

JLimperg committed:
Add dependency_set_of_hyp_inclusive and variants

robertylewis reviewed (commented):
> @robertylewis could you take another look? Sorry about the duplicate work.

For future reference, please remove the `awaiting-review` tag and add `WIP` or `awaiting-author` if you have changes planned!

I see you're making edits now. Ping me when you're done and I'll review again.

JLimperg committed:
Add doc string

robertylewis commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,501 @@
+
+namespace tactic
+
+private meta def local_list_to_name_set (lcs : list expr) : name_set :=
```
This could move elsewhere and become public.

robertylewis commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,501 @@
+private meta def local_list_to_name_set (lcs : list expr) : name_set :=
+lcs.foldl (λ ns h, ns.insert h.local_uniq_name) mk_name_set
+
+/-! ## Direct Dependencies -/
```
```suggestion
/-! ### Direct Dependencies -/
```

robertylewis commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,501 @@
+hyp_directly_depends_on_local_name_set h $ local_list_to_name_set hs
+
+
+/-! ## (Indirect) Dependencies -/
```
```suggestion
/-! ### (Indirect) Dependencies -/
```
I think the word "transitive" should appear somewhere around here, since that would be my first thought to ctrl-f.

robertylewis commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,501 @@
+
+/--
+`hyp_depends_on_local_set h hs` is true iff the hypothesis `h` depends on
+any of the hypotheses `hs`. See `tactic.hyp_depends_on_local_name_set`.
```
Perhaps repeat the same "this is expensive" warning instead of the reference?

robertylewis commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,501 @@
+rb_set.to_list <$> dependency_set_of_hyp_inclusive h
+
+
+/-! ## Reverse Dependencies -/
```
```suggestion
/-! ### Reverse Dependencies -/
```

robertylewis commented on src/tactic/dependencies.lean:
```diff
@@ -0,0 +1,501 @@
+  tactic (list expr) :=
+reverse_dependencies_of_hyp_name_set_inclusive $ local_list_to_name_set hs
+
+/-! ## Reverting -/
```
```suggestion
/-! ### Reverting -/
```

robertylewis reviewed (commented):
This looks useful and exhaustive. Thank you for filling out the full API here!

I have no major concerns. AFAIK all transitive checks go through `context_dependencies`, which computes the `expr -> expr_set` map for the entire context, right? It's not clear to me that doing this greedily is always ideal. There are certainly applications where keeping a graph representation and computing sets on demand would be more efficient, right? But presumably this approach is better at other times. I'm not sure how much to worry about this.

JLimperg committed:
Apply suggestions from code review

Co-authored-by: Rob Lewis <Rob.y.lewis@gmail.com>

JLimperg committed:
Make local_list_to_name_set public

JLimperg committed:
Make performance warning more explicit

JLimperg committed:
Add 'transitive' to heading for grepability

JLimperg committed:
Reimplement dependency search; add functions to get dependencies of multiple hyps; improve docs
